src/layer-composer/types.ts

```diff
@@ -0,0 +1,65 @@
+import { AnySourceImpl, Layer } from 'mapbox-gl'
+import { ColorRamps } from 'layer-composer/generators/heatmap/heatmap'
+
+export interface LayerComposerGl {
+  id: string
+  sources: AnySourceImpl[]
+  layers: Layer[]
+  promise?: Promise<LayerComposerGl>
+  promises?: Promise<LayerComposerGl>[]
+}
+
+export interface LayerComposerGenerator {
+  type: string
+  getStyle: (layer: LayerComposerLayer) => LayerComposerGl
+}
+
+// TODO FIRST: DEFINE THIS
+export interface LayerComposeStyles {
+  style: any
+  promises?: Promise<any>[]
+}
+
+export interface LayerComposerOptions {
+  generators?: { [key: string]: LayerComposerGenerator }
+  version?: number
+  glyphs?: string
+  sprite?: string
+}
+
+export interface LayerComposerLayer {
```

Comment:
It seems awkward to me to have all generator params as a single type, with almost all parameters optional. This defeats the purpose of typing, and most of those parameters are actually mandatory when they are used with their specific Generator. But maybe you meant to split by generator type later ? 

I think we can have an object implementing multiple interfaces, which mean we would have:
```
interface GeneratorConfig {
  id: string
  type: 'BACKGROUND' | 'BASEMAP' | 'CARTO_POLYGONS' | 'GL_STYLES' | 'HEATMAP' | string
}
interface BackgroundGeneratorConfig {
  color: string
}
```
...
```
class Config implements GeneratorConfig, BackgroundGeneratorConfig
class BackgroundGenerator {
  getStyle = (config: Config) => {
```

What do you think?